README
             UPPRINT - tools for better n-up printing

                                * * *

DESCRIPTION

 The upprint package contains tools for better formatting of print
 jobs. Unlike print systems such as CUPS, which delegate all
 formatting to the print server, the upprint tools are designed to
 work on the user side, and therefore under the user's control.
 The package provides the following tools:

 * pstops-clip: an extension of Angus J. C. Duggan's "pstops" program
   for reformatting postscript files. The extension allows
   configurable page clipping.

 * psdim: a small utility to be used in conjunction with pstops.  It
   determines the true margins of a postscript documents and
   calculates the optimal page placement for n-up printing. It outputs
   a format string suitable for processing by pstops or pstops-clip.
   In order to accurately determine the size of the pages in a
   postscript document, psdim internally uses the ghostscript
   interpreter to render the document. For this reason, psdim is
   relatively slow, but much more accurate than other tools.

 * lprwrap: a wrapper around the standard "lpr" program that performs
   client-side pre-processing of print jobs. It recognizes special
   options for duplex printing and n-up printing, and can be used as a
   drop-in replacement for "lpr".

DEPENDENCIES

 For psdim to work, ghostscript must be installed on your system.

 For lprwrap to work, you need at least: bash, lpr, file, mktemp,
 sed, and awk. For enhanced functionality, you should also have:
 getopt, mpage, gzip, bzip2, iconv, dvips, and pdf2ps.

COMPILATION AND INSTALLATION

 For generic compilation instructions, see the file INSTALL. In
 practice the following three commands to the trick:

 ./configure
 make
 make install

 The ./configure script recognizes the following non-standard options:
   --enable-metric         use metric units (centimeters) as default
   --enable-a4             use a4 as the default papersize

CUSTOMIZATION

 lprwrap is designed as a drop-in wrapper around the "lpr"
 program. After installing lprwrap, you can create an alias, for
 example by putting the following into your .bashrc or
 the system-wide /etc/bash.bashrc:

    alias lpr='lprwrap -h'

USAGE

 See the respective man pages for full usage information.

 PSDIM USAGE:

  Usage: psdim [options] [file]
  
  Options:
   -h, --help               - print this help message and exit
   -v, --version            - print version info and exit
   -l, --license            - print license info and exit
   -q, --quiet              - suppress progress info on stderr
   -x, --width <dim>        - output page width
   -y, --height <dim>       - output page height
   -p, --page <format>      - select output page size (e.g. letter, a4)
   -m, --margin <dim>       - margin around page
   -n, --hmargin <dim>      - horizontal margin (horizontal space)
   -o, --vmargin <dim>      - vertical margin (vertical space)
   -s, --sep <dim>          - separation between pages
   -t, --hsep <dim>         - horizontal separation
   -u, --vsep <dim>         - vertical separation
   -L, --landscape          - landscape orientation
   -R, --seascape           - seascape orientation
   -U, --upside-down        - upside down orientation
   -P, --portrait           - portrait orientation
   -f, --format <n>x<m>     - arrange pages in n rows and m columns
   -a, --hpolicy <n>        - horizontal alignment policy
   -b, --vpolicy <n>        - vertical alignment policy
   -c, --columnmode         - sort pages in columns
   -d, --righttoleft        - page numbers increase right to left
   -e, --bottomtotop        - page numbers increase bottom to top
   -C, --color              - handle non-white backgrounds
   -i, --clip               - output page clipping instructions for pstops-clip
   -F, --fudge [L|R|T|B]<n> - percentage of pixels allowed out of bounds
   -H, --ladjust <dim>      - adjust left edge of input's bounding box
   -I, --radjust <dim>      - adjust right edge of input's bounding box
   -J, --tadjust <dim>      - adjust top edge of input's bounding box
   -K, --badjust <dim>      - adjust bottom edge of input's bounding box
   -S, --shrink             - only shrink, never enlarge page
   -1, --1up                - fit to size mode
   -2, --2up                - 1x2 landscape mode
   -4, --4up                - 2x2 portrait mode
   -8, --8up                - 2x4 landscape mode
   -9, --9up                - 3x3 portrait mode
   -6, --16up               - 4x4 portrait mode
  
  Dimensions can have optional units, e.g. 6.5in, 15cm. Default is
  inches unless --enable-metric was configured during compilation.

  Valid page formats are: a4, a3, a5, b5, letter, legal, tabloid,
  statement, executive, folio, quarto, 10x14. Default is letter unless
  --enable-a4 was configured during compilation.

 PSTOPS-CLIP USAGE:

  Usage: pstops-clip [options] <pagespecs> [infile [outfile]]
  Options:
   --help               - print this help message and exit
   --version, -v        - print version info and exit
   --license            - print license info and exit
   -q                   - don't print page numbers to stderr
   -b                   - prevent PostScript 'bind' operators from binding
   -w<width>            - set page width
   -h<height>           - set page height
   -p<paper>            - set paper size
   -d[linewidth]        - draw a box around each page
  
  Paper sizes:
   a3, a4, a5, b5, letter, legal, tabloid, statement, executive, folio,
   quarto, 10x14, _glibc
  
  Page specification:
   <pagespecs>  = [modulo:]<spec>[,spec|+spec]...
   <spec>       = [-]pageno[@scale][L|R|U][(xoff,yoff)][{x0,y0,x1,y1}]
   modulo:      the number of pages in each block (1 or more; default 1)
   spec:        the page specification for one page in each block
   pageno:      between 0 and modulo-1
   scale:       scaling factor
   L,R,U:       rotate left, right, upside-down
   xoff,yoff:   page origin in PostScript points, or units of cm, in, w, h
   x0,y1,x1,y1: page clip path in PostScript points, or units of cm, in, w, h

 LPRWRAP USAGE:

  Usage: lprwrap [options] [file]...
  
  Options: same options as for lpr, plus:
    --help             print usage info and exit
    --version          print version info and exit
    --license          print license info and exit
    --verbose          print some diagnostics to stderr
    --test             send output to stdout instead of printer
    --force            force suspicious files to be printed anyway
    --tmpdir <dir>     choose directory for temporary files
    --lpr <name>       program to use for the backend lpr
    -oduplex, -od      duplex printing (with long edge binding)
    -ohduplex, -oh     duplex printing (with short edge binding)
    -osimplex, -os     no duplex printing (default)
    -o<paper>          select a paper size: letter, executive, legal, a3,
                       a4, a5, b5, tabloid, statement, folio, quarto, 10x14
                       (default is letter)
    -o<tray>           select a tray: upper, middle, lower, manual
    -o2up              2-up printing
    -o4up              4-up printing
    -o8up              8-up printing
    -o9up              9-up printing
    -o16up             16-up printing
    -odumb             faster, but uglier n-up printing (don't fit to size)
    -ofit              fit to size in 1-up mode
    -osolid            detect non-white background colors (slower)
    -ofudge=[LRTB]<n>  percentage of pixels allowed out of bounds
    -obox              print a box around each page (plain text only)
    -oinfo             print date and user name on each page (plain text only)
    -ouncollated       with -#, make NUM copies of each page, not of each file
                       (may not work in duplex mode)
    -ocups             let CUPS handle duplexing, paper and tray selection, 
                       and collation (by default the frontend handles it)
  
  Note that -duplex and -hduplex refer to whether you turn the page on
  the long edge or on the short edge, with respect to the *input*
  page. That is, if you select -o2up, you should still select -oduplex,
  just as you would if you hadn't selected -o2up.  But if your input
  file is in landscape format, you should probably select -ohduplex.
  
  The following standard lpr options are also supported:
  
    -P <printer>       direct output to a specific printer
    -h                 do not print burst page
    -m                 send mail upon completion
    -r                 remove the file upon completion of spooling
    -# <num>           number of copies to make of each file
    -[1234] <font>     specify a font to be mounted on font position i
    -C <class>         job classification to use on the burst page
    -J <job>           job name to print on the burst page
    -U <user>          user name to print on the burst page and for accounting.
                       This option is only honored if the real user-id
                       is daemon (or that specified in the printcap file
                       instead of daemon), and is intended for those
                       instances where print filters wish to requeue
                       jobs.
  
  The following options are ignored for backwards compatibility:
  
    -T <title>, -i <numcols>, -w <num>, -c, -d, -f, -g, -l, -n, -p, -t, -v, -s
  
  Command line options are also read from the two files /etc/lprrc and
  $HOME/.lprrc, if they exist, and processed in that order before any
  other command line options are processed. For instance, if you like
  duplex printing to be your default, put "-od" into one of these
  files, or if you would like to use a private temp directory, put
  "--tmpdir <dir>". If your system uses CUPS, put "-ocups". You can
  also use this to set the default paper size.

COPYRIGHT

 For psdim and lprwrap:
 ---------------------- 

 Copyright (C) 2001-2012 Peter Selinger

 This program is free software; you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation; either version 2 of the License, or (at
 your option) any later version.

 This program is distributed in the hope that it will be useful, but
 WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 General Public License for more details.

 You should have received a copy of the GNU General Public License
 along with this program; if not, write to the Free Software
 Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307,
 USA. See also http://www.gnu.org/.

 See the file COPYING for details.

 For pstops-clip:
 ----------------

 Copyright (C) 1991-1995 Angus J. C. Duggan.
 Copyright (C) 2006-2012 Peter Selinger.

 This program may be copied and used for any purpose (including
 distribution as part of a for-profit product), provided:

 1) The original attribution of the programs is clearly displayed in
    the product and/or documentation, even if the programs are modified
    and/or renamed as part of the product.

 2) The original source code of the programs is provided free of
    charge (except for reasonable distribution costs). For a
    definition of reasonable distribution costs, see the Gnu General
    Public License or Larry Wall's Artistic License (provided with the
    Perl 4 kit). The GPL and Artistic License in NO WAY affect this
    license; they are merely used as examples of the spirit in which
    it is intended.

 3) These programs are provided "as-is". No warranty or guarantee of
    their fitness for any particular task is provided. Use of these
    programs is completely at your own risk.

 See the file pstops-clip/LICENSE for details.


NEWS
March 27, 2012: upprint, release 1.7

 Portability improvements. Lprwrap: pass unrecognized -o options to
 the underlying lpr, so that applications expecting a CUPS lpr will
 still work. Changed the way lprwrap calls the "real" lpr.

July 23, 2011: upprint, release 1.6

 Improved portability. Lprwrap: Compatibility with newer versions of
 CUPS has been improved. The -ocups option to lprwrap now replaces the
 old lprwrap-cups script. Allow multiple -ofudge options. New --lpr
 option can be used to specify the backend "lpr" executable. The
 TEMPDIR environment variable is now respected. Support for pdftops
 added.  Added a size limit on text files, and a --force option to
 override it. Psdim: when compiled with metric defaults, use metric
 units in the output format string. New options --[lrtb]adjust for
 adjusting the bounding box of input document. Compensate for certain
 PostScript files generated by Acrobat Reader that resulted in shifted
 output.

February 15, 2007: upprint, release 1.5

 The new upprint package combines the features of the old lpr-wrapper
 and psdim packages. It contains lprwrap, psdim, as well as the new
 program pstops-clip. Page clipping was improved, so that documents
 with opaque backgrounds can now be printed in n-up mode. A 'fudge'
 feature was added to allow more flexible margin detection.  The
 lpr-wrapper program was renamed lprwrap.

January 22, 2007: lpr-wrapper, release 0.7

 Better CUPS support was added. PPD printer description files are no
 longer required. Printing of Unicode text files is now supported via
 iconv.

March 30, 2005: psdim, release 1.4

 This is a maintenance release. The sources have been updated to
 compile with recent versions of gcc.

January 15, 2004: lpr-wrapper, release 0.6

 We now treat any unknown file type consisting of printable characters
 as text. This allows printing text-based files such as GPG keys,
 which are not identified as "text" by the "file" program. We also
 call the backend "lpr" with the "-l" option to prevent filtering in
 CUPS. 

December 6, 2003: psdim, release 1.3

 This release does not differ from previous releases in functionality.
 The sources have been restructured to take advantage of standard
 automake/autoconf configuration. Configuration options --enable-a4
 and --enable-metric have been added. A spurious source of "Postscript
 errors" has been eliminated.


psdim/main.c
/* Copyright (C) 2001-2012 Peter Selinger.
   This file is part of the upprint package. It is free software and
   is distributed under the terms of the GNU general public license.
   See the file COPYING for details. */

/* $Id: main.c 100 2012-03-27 23:02:39Z selinger $ */

/* psdim: a small utility to be used in conjunction with pstops.  It
   uses GhostScript to determine the "actual" size of the printed
   pages in a document. It then figures out the optimal arrangement of
   pages in a grid for 2up printing or other such printing formats.
   The output is a format specification which can be passed to
   pstops. */

/* Requirements: ghostscript ("gs") must be installed. */

#ifdef HAVE_CONFIG_H
 #include "config.h"
#endif

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>
#include <getopt.h>

#include "main.h"
#include "psdim.h"
#include "format.h"

info_t info;

/* define a set of error conditions, in the style of errno. */

int merrno;  /* for passing error conditions */

char *mstrerror[] = {
  /* 0 */    NULL,
  /* 1 */    "Out of memory", 
  /* 2 */    "Could not invoke ghostscript",
  /* 3 */    "Unexpected end of file from ghostscript",
  /* 4 */    "Postscript error",
  /* 5 */    "I/O error",  /* errno will be set */
};

/* dimensions of the various page formats, in postscript points */
pageformat_t pageformat[] = {
  { "a4",	 595,  842 },
  { "a3",        842, 1191 },
  { "a5",        421,  595 },
  { "b5",        516,  729 },
  { "letter",    612,  792 },
  { "legal",     612, 1008 },
  { "tabloid",   792, 1224 },
  { "statement", 396,  612 },
  { "executive", 540,  720 },
  { "folio",     612,  936 },
  { "quarto",    610,  780 },
  { "10x14",     720, 1008 },
  { NULL, 0, 0 },
};

#define MASK_LEFT   0x1
#define MASK_RIGHT  0x2
#define MASK_BOTTOM 0x4
#define MASK_TOP    0x8

int license(FILE *f) {
  fprintf(f, 
"This program is free software; you can redistribute it and/or modify\n"
"it under the terms of the GNU General Public License as published by\n"
"the Free Software Foundation; either version 2 of the License, or\n"
"(at your option) any later version.\n"
"\n"
"This program is distributed in the hope that it will be useful,\n"
"but WITHOUT ANY WARRANTY; without even the implied warranty of\n"
"MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the\n"
"GNU General Public License for more details.\n"
"\n"
"You should have received a copy of the GNU General Public License\n"
"along with this program; if not, write to the Free Software\n"
"Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307, USA.\n"
);

  return 0;

}

int usage(FILE *f) {
  int i, j;

  fprintf(f, "Usage: "PSDIM" [options] [file]\n\n");
  fprintf(f, "Options:\n");
  fprintf(f, " -h, --help               - print this help message and exit\n");
  fprintf(f, " -v, --version            - print version info and exit\n");
  fprintf(f, " -l, --license            - print license info and exit\n");
  fprintf(f, " -q, --quiet              - suppress progress info on stderr\n");
  fprintf(f, " -x, --width <dim>        - output page width\n");
  fprintf(f, " -y, --height <dim>       - output page height\n");
  fprintf(f, " -p, --page <format>      - select output page size (e.g. letter, a4)\n");
  fprintf(f, " -m, --margin <dim>       - margin around page\n");
  fprintf(f, " -n, --hmargin <dim>      - horizontal margin (horizontal space)\n");
  fprintf(f, " -o, --vmargin <dim>      - vertical margin (vertical space)\n");
  fprintf(f, " -s, --sep <dim>          - separation between pages\n");
  fprintf(f, " -t, --hsep <dim>         - horizontal separation\n");
  fprintf(f, " -u, --vsep <dim>         - vertical separation\n");
  fprintf(f, " -L, --landscape          - landscape orientation\n");
  fprintf(f, " -R, --seascape           - seascape orientation\n");
  fprintf(f, " -U, --upside-down        - upside down orientation\n");
  fprintf(f, " -P, --portrait           - portrait orientation\n");
  fprintf(f, " -f, --format <n>x<m>     - arrange pages in n rows and m columns\n");
  fprintf(f, " -a, --hpolicy <n>        - horizontal alignment policy\n");
  fprintf(f, " -b, --vpolicy <n>        - vertical alignment policy\n");
  fprintf(f, " -c, --columnmode         - sort pages in columns\n");
  fprintf(f, " -d, --righttoleft        - page numbers increase right to left\n");
  fprintf(f, " -e, --bottomtotop        - page numbers increase bottom to top\n");
  fprintf(f, " -C, --color              - handle non-white backgrounds\n");
  fprintf(f, " -i, --clip               - output page clipping instructions for pstops-clip\n");
  fprintf(f, " -F, --fudge [L|R|T|B]<n> - percentage of pixels allowed out of bounds\n");
  fprintf(f, " -H, --ladjust <dim>      - adjust left edge of input's bounding box\n");
  fprintf(f, " -I, --radjust <dim>      - adjust right edge of input's bounding box\n");
  fprintf(f, " -J, --tadjust <dim>      - adjust top edge of input's bounding box\n");
  fprintf(f, " -K, --badjust <dim>      - adjust bottom edge of input's bounding box\n");
  fprintf(f, " -S, --shrink             - only shrink, never enlarge page\n");
  fprintf(f, " -1, --1up                - fit to size mode\n");
  fprintf(f, " -2, --2up                - 1x2 landscape mode\n");
  fprintf(f, " -4, --4up                - 2x2 portrait mode\n");
  fprintf(f, " -8, --8up                - 2x4 landscape mode\n");
  fprintf(f, " -9, --9up                - 3x3 portrait mode\n");
  fprintf(f, " -6, --16up               - 4x4 portrait mode\n");
  fprintf(f, "\n");
  fprintf(f, "Dimensions can have optional units, e.g. 6.5in, 15cm (default: "DEFAULT_UNIT_NAME").\n\n");

  j = fprintf(f, "Valid page formats are: ");
  for (i=0; pageformat[i].name!=NULL; i++) {
    if (j + strlen(pageformat[i].name) > 75) {
      fprintf(f, "\n");
      j = 0;
    }
    j += fprintf(f, "%s, ", pageformat[i].name);
  }
  fprintf(f, "(default: "
#ifdef USE_A4
	  "a4"
#else
	  "letter"
#endif
	  ").\n");
  
  return 0;
}

/* parse a dimension of the kind "1.5in", "7cm", etc. Return result in
   postscript points (=1/72 in). If endptr!=NULL, store pointer to
   next character in *endptr in the manner of strtod(3). */
double parse_dimension(char *s, char **endptr) {
  char *p;
  double x;

  x = strtod(s, &p);
  if (!*p) {
    x *= DEFAULT_UNIT_POINTS;
  } else if (!strcasecmp(p, "in")) {
    x *= 72;
    p += 2;
  } else if (!strcasecmp(p, "cm")) {
    x *= 72 / 2.54;
    p += 2;
  } else if (!strcasecmp(p, "mm")) {
    x *= 72 / 25.4;
    p += 2;
  } else if (!strcasecmp(p, "pt")) {
    p += 2;
  }
  if (endptr!=NULL) {
    *endptr = p;
  }
  return x;
}

static struct option longopts[] = {
  {"help",         0, 0, 'h'},
  {"version",      0, 0, 'v'},
  {"license",      0, 0, 'l'},
  {"quiet",        0, 0, 'q'},
  {"width",        1, 0, 'x'},
  {"height",       1, 0, 'y'},
  {"page",         1, 0, 'p'},
  {"margin",       1, 0, 'm'},
  {"hmargin",      1, 0, 'n'},
  {"vmargin",      1, 0, 'o'},
  {"sep",          1, 0, 's'},
  {"hsep",         1, 0, 't'},
  {"vsep",         1, 0, 'u'},
  {"ladjust",      1, 0, 'H'},
  {"radjust",      1, 0, 'I'},
  {"tadjust",      1, 0, 'J'},
  {"badjust",      1, 0, 'K'},
  {"landscape",    0, 0, 'L'},
  {"seascape",     0, 0, 'R'},
  {"upside-down",  0, 0, 'U'},
  {"portrait",     0, 0, 'P'},
  {"format",       1, 0, 'f'},
  {"hpolicy",      1, 0, 'a'},
  {"vpolicy",      1, 0, 'b'},
  {"columnmode",   0, 0, 'c'},
  {"righttoleft",  0, 0, 'd'},
  {"bottomtotop",  0, 0, 'e'},
  {"color",        0, 0, 'C'},
  {"clip",         0, 0, 'i'},
  {"fudge",        1, 0, 'F'},
  {"shrink",       0, 0, 'S'},
  {"1up",          0, 0, '1'},
  {"2up",          0, 0, '2'},
  {"4up",          0, 0, '4'},
  {"8up",          0, 0, '8'},
  {"9up",          0, 0, '9'},
  {"16up",         0, 0, '6'},
  {0, 0, 0, 0}
};

static char *shortopts = "hvlqx:y:p:m:n:o:s:t:u:LRUPf:a:b:cdeCiF:S124896H:I:J:K:";

int dopts(int ac, char *av[]) {
  int c, i, j;
  char *p;
  double fudge;
  int mask;

  /* defaults */
#ifdef USE_A4
  info.w = 595;       /* default page is a4 */
  info.h = 842;
#else
  info.w = 612;       /* default page is letter */
  info.h = 792;
#endif
  info.hmargin = 36;  /* default margin 0.5in */
  info.vmargin = 36;  /* default margin 0.5in */
  info.hsep = 36;     /* default sep 0.5in */
  info.vsep = 36;     /* default sep 0.5in */
  info.ladjust = 0;   /* default adjustment 0in */
  info.radjust = 0;   /* default adjustment 0in */
  info.tadjust = 0;   /* default adjustment 0in */
  info.badjust = 0;   /* default adjustment 0in */

  info.land = 1;      /* default: 2up */
  info.cols = 2;
  info.rows = 1;
  info.columnmode = 0;
  info.righttoleft = 0; 
  info.bottomtotop = 0; 
  info.hpolicy = 1;
  info.vpolicy = 1;
  info.quiet = 0;
  info.infile = NULL;
  info.percentile.x0 = 0.0;
  info.percentile.x1 = 1.0;
  info.percentile.y0 = 0.0;
  info.percentile.y1 = 1.0;

  info.color = 0;
  info.clip = 0;

  while ((c = getopt_long(ac, av, shortopts, longopts, NULL)) != -1) {
    switch (c) {
    case 'h':
      fprintf(stdout, ""PSDIM" "VERSION". Determines optimal page format for pstops.\n\n");
      usage(stdout);
      exit(0);
      break;
    case 'v':
      fprintf(stdout, ""PSDIM" "VERSION". Copyright (C) 2001-2012 Peter Selinger.\n");
      exit(0);
      break;
    case 'l':
      fprintf(stdout, ""PSDIM" "VERSION". Copyright (C) 2001-2012 Peter Selinger.\n\n");
      license(stdout);
      exit(0);
      break;
    case 'q':
      info.quiet = 1;
      break;
    case 'x':
      info.w = parse_dimension(optarg, &p);
      if (*p) {
	fprintf(stderr, ""PSDIM": invalid dimension -- %s\n", optarg);
	exit(1);
      }
      break;
    case 'y':
      info.h = parse_dimension(optarg, &p);
      if (*p) {
	fprintf(stderr, ""PSDIM": invalid dimension -- %s\n", optarg);
	exit(1);
      }
      break;
    case 'p':
      for (i=0; pageformat[i].name!=NULL; i++) {
	if (strcasecmp(pageformat[i].name, optarg)==0) 
	  break;
      }
      if (!pageformat[i].name) {
	fprintf(stderr, ""PSDIM": unrecognized page format -- %s\n", optarg);
	j = fprintf(stderr, "Use one of: ");
	for (i=0; pageformat[i].name!=NULL; i++) {
	  if (j + strlen(pageformat[i].name) > 75) {
	    fprintf(stderr, "\n");
	    j = 0;
	  }
	  j += fprintf(stderr, "%s, ", pageformat[i].name);
	}
	fprintf(stderr, "or use -x and -y options.\n");
	exit(1);
      }  
      info.w = pageformat[i].w;
      info.h = pageformat[i].h;
      break;
    case 'm':
      info.hmargin = info.vmargin = parse_dimension(optarg, &p);
      if (*p) {
	fprintf(stderr, ""PSDIM": invalid dimension -- %s\n", optarg);
	exit(1);
      }
      break;
    case 'n':
      info.hmargin = parse_dimension(optarg, &p);
      if (*p) {
	fprintf(stderr, ""PSDIM": invalid dimension -- %s\n", optarg);
	exit(1);
      }
      break;
    case 'o':
      info.vmargin = parse_dimension(optarg, &p);
      if (*p) {
	fprintf(stderr, ""PSDIM": invalid dimension -- %s\n", optarg);
	exit(1);
      }
      break;
    case 's':
      info.hsep = info.vsep = parse_dimension(optarg, &p);
      if (*p) {
	fprintf(stderr, ""PSDIM": invalid dimension -- %s\n", optarg);
	exit(1);
      }
      break;
    case 't':
      info.hsep = parse_dimension(optarg, &p);
      if (*p) {
	fprintf(stderr, ""PSDIM": invalid dimension -- %s\n", optarg);
	exit(1);
      }
      break;
    case 'u':
      info.vsep = parse_dimension(optarg, &p);
      if (*p) {
	fprintf(stderr, ""PSDIM": invalid dimension -- %s\n", optarg);
	exit(1);
      }
      break;
    case 'H':
      info.ladjust = parse_dimension(optarg, &p);
      if (*p) {
	fprintf(stderr, ""PSDIM": invalid dimension -- %s\n", optarg);
	exit(1);
      }
      break;
    case 'I':
      info.radjust = parse_dimension(optarg, &p);
      if (*p) {
	fprintf(stderr, ""PSDIM": invalid dimension -- %s\n", optarg);
	exit(1);
      }
      break;
    case 'J':
      info.tadjust = parse_dimension(optarg, &p);
      if (*p) {
	fprintf(stderr, ""PSDIM": invalid dimension -- %s\n", optarg);
	exit(1);
      }
      break;
    case 'K':
      info.badjust = parse_dimension(optarg, &p);
      if (*p) {
	fprintf(stderr, ""PSDIM": invalid dimension -- %s\n", optarg);
	exit(1);
      }
      break;
    case 'L':
      info.land = 1;
      break;
    case 'R':
      info.land = 3;
      break;
    case 'U':
      info.land = 2;
      break;
    case 'P':
      info.land = 0;
      break;
    case 'f':
      info.rows = strtol(optarg, &p, 10);
      if (*p == 'x' && info.rows>0) {
	info.cols = strtol(p+1, &p, 10);
	if (*p == '\0' && info.cols>0) {
	  break;
	}
      }
      fprintf(stderr, ""PSDIM": invalid format -- %s\n", optarg);
      fprintf(stderr, "Use <n>x<m>.\n");
      exit(1);
    case 'a':
      info.hpolicy = atoi(optarg);
      if (info.hpolicy<0 || info.hpolicy>4) {
	fprintf(stderr, ""PSDIM": invalid horizontal alignment policy -- %d\n", info.hpolicy);
	fprintf(stderr, "Valid range is 0--4.\n");
	exit(1);
      }
      break;
    case 'b':
      info.vpolicy = atoi(optarg);
      if (info.vpolicy<0 || info.vpolicy>4) {
	fprintf(stderr, ""PSDIM": invalid vertical alignment policy -- %d\n", info.vpolicy);
	fprintf(stderr, "Valid range is 0--4.\n");
	exit(1);
      }
      break;
    case 'c':
      info.columnmode = 1;
      break;
    case 'd':
      info.righttoleft = 1;
      break;
    case 'e':
      info.bottomtotop = 1;
      break;
    case 'C':
      info.color = 1;
      break;
    case 'i':
      info.clip = 1;
      break;
    case 'F': {
      p = optarg;
      mask = 0;
      while (1) {
	if (*p == 'l' || *p == 'L') {
	  mask |= MASK_LEFT;
	} else if (*p == 'r' || *p == 'R') {
	  mask |= MASK_RIGHT;
	} else if (*p == 'b' || *p == 'B' || *p == 'd' || *p == 'D') {
	  mask |= MASK_BOTTOM;
	} else if (*p == 't' || *p == 'T' || *p == 'u' || *p == 'U') {
	  mask |= MASK_TOP;
	} else {
	  break;
	}
	p += 1;
      }
      fudge = strtod(p, &p);
      if (*p || fudge < 0.0 || fudge > 100.0) {
	fprintf(stderr, ""PSDIM": invalid percentage -- %s\n", optarg);
	exit(1);
      }
      if (mask == 0 || mask & MASK_LEFT) {
	info.percentile.x0 = fudge * 0.01;
      }
      if (mask == 0 || mask & MASK_RIGHT) {
	info.percentile.x1 = 1.0 - fudge * 0.01;
      }
      if (mask == 0 || mask & MASK_BOTTOM) {
	info.percentile.y0 = fudge * 0.01;
      }
      if (mask == 0 || mask & MASK_TOP) {
	info.percentile.y1 = 1.0 - fudge * 0.01;
      }
      break;
    }
    case 'S':
      info.shrink = 1;
      break;
    case '1':
      info.land = 0;
      info.cols = 1;
      info.rows = 1;
      break;
    case '2':
      info.land = 1;
      info.cols = 2;
      info.rows = 1;
      break;
    case '4':
      info.land = 0;
      info.cols = 2;
      info.rows = 2;
      break;
    case '8':
      info.land = 1;
      info.cols = 4;
      info.rows = 2;
      break;
    case '9':
      info.land = 0;
      info.cols = 3;
      info.rows = 3;
      break;
    case '6':
      info.land = 0;
      info.cols = 4;
      info.rows = 4;
      break;
    case '?':
      fprintf(stderr, "Try --help for more info\n");
      exit(1);
      break;
    default:
      fprintf(stderr, ""PSDIM": Invalid option -- %c\n", c);
      exit(1);
    }
  }

  if (optind < ac) {
    info.infile = av[optind];
    optind++;
  }
  if (optind < ac) {
    fprintf(stderr, ""PSDIM": Too many non-option arguments\n");
    fprintf(stderr, "Try --help for more info\n");
    exit(1);
  }

  return 0;
}

void adjust(info_t info, int n, bbox_t *bboxes) {
  int i;

  for ( i=0; i<n; i++) {
#ifdef DEBUG
      fprintf(stderr, "### bbox(%d) was: x0=%d x1=%d y0=%d, y1=%d\n",
	      i, bboxes[i].x0, bboxes[i].x1, bboxes[i].y0, bboxes[i].y1 );
#endif
    bboxes[i].x0 += info.ladjust;
    bboxes[i].x1 -= info.radjust;
    bboxes[i].y0 += info.badjust;
    bboxes[i].y1 -= info.tadjust;
#ifdef DEBUG
      fprintf(stderr, "### bbox(%d) now: x0=%d x1=%d y0=%d, y1=%d\n",
	      i, bboxes[i].x0, bboxes[i].x1, bboxes[i].y0, bboxes[i].y1 );
#endif
  }
}
  
int main(int ac, char *av[]) {
  int n;
  int r;
  bbox_t *bboxes;

  /* read command line options */
  dopts(ac, av);
  
  n = info.rows * info.cols;

  /* allocate bounding boxes for n sets of pages */
  bboxes = (bbox_t *)malloc(n*sizeof(bbox_t));
  if (!bboxes) {
    fprintf(stderr, ""PSDIM": Out of memory\n");
    exit(1);
  }
  
  /* extract bounding boxes from file */
  if (info.color) {
    r = psdim_color(info.infile, n, bboxes, &info.percentile);
  } else {
    r = psdim(info.infile, n, bboxes, &info.percentile);
  }
  if (r == -1) {
    if (merrno != ME_IO) {
      fprintf(stderr, ""PSDIM": %s\n", mstrerror[merrno]);
    } else {
      fprintf(stderr, ""PSDIM": %s\n", strerror(errno));
    }      
    return merrno;
  }

  /* apply additional adjustment */
  adjust(info, n, bboxes);

  /* calculate and output best pstops format */
  format(info, n, bboxes);

  return 0;
}

